### PR TITLE
Adds initial support for better retrieving

### PR DESCRIPTION
**Pull request by lukecampbell**
All the changes are purely additive and should not affect any existing function.

Adds a new Complex Coverage type: TIMESERIES
TIMESERIES has the ability to do random inserts and random reads with good performance.

Adds a new function to all coverage models: `get_value_dictionary`
Behaves just like `get_parameter_values` except that it returns a dictionary of values, it can slice on the temporal values as well as slice on the domain values and it doesn't require the entire temporal domain to verify slices. A huge improvement for retrieval operations, now we won't have to look at the entire "time" array every time we make a retrieve operation. 

The value dictionaries are sorted IF any subset of the coverages overlap, and duplicates are removed.

Part of [OOIION-1310](https://jira.oceanobservatories.org/tasks/browse/OOIION-1310)

[Covmodel_test](http://buildbot.oceanobservatories.org:8010/builders/covmodel_test/builds/144)

### CODE REVIEW

**Comment by lukecampbell** [comment]
@jamie-cyber1 Tests pass, ready for review

File: coverage_model/coverage.py [hunk 1]
```diff
@@ -404,6 +406,95 @@ def get_parameter_values(self, param_name, tdoa=None, sdoa=None, return_value=No
 
         return return_value
 
+    def get_value_dictionary(self, param_list=None, temporal_slice=None, domain_slice=None):
+        '''
+        Retrieves a dictionary of parameters and value sets.
+        The value set is a cross-section result between the specified temporal_slice
+        '''
+        if self.closed:
+            raise IOError('I/O operation on closed file')
```
Comment (lukecampbell) [inline coverage_model/coverage.py:415]: What's this do?

File: coverage_model/coverage.py [hunk 2]
```diff
@@ -404,6 +406,95 @@ def get_parameter_values(self, param_name, tdoa=None, sdoa=None, return_value=No
...
+        # But remember to remove it on the way out
+        time_vector = self._range_value[self.temporal_parameter_name]
+
+        # Dramatically improve performance, no need to retrieve the entire
+        # time-array to find out where start and stop are, we just use a
+        # binary-search on an "array-like" HDF5 interface
+
+        start_idx = np.searchsorted(time_vector, start)
+        stop_idx = np.searchsorted(time_vector, stop)
+        slice_ = slice(start_idx, stop_idx, stride)
```
Comment (lukecampbell) [inline coverage_model/coverage.py:485]: I think you misspelled `slice_`

**Comment by jamie-cyber1** [comment]
:+1:

**Pull request merged**